### PR TITLE
Merge 'feature/cloud9'

### PR DESCRIPTION
**Pull request by justinmk3**
## Description

Add support for running in Cloud9.

## Test cases

-  VSIX artifact contains the correct quickstart/marketplace page
    - Links work in the quickstart 
- SAM debug with Image resource in cloud9
    - Image+python 3.8 issue: https://github.com/aws/aws-toolkit-vscode/issues/1442
    - Image+API+nodejs issue: https://github.com/aws/aws-toolkit-vscode/issues/1443

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/credentials/credentialsUtilities.ts [hunk 1]
```diff
@@ -27,6 +28,8 @@ export function asEnvironmentVariables(credentials: Credentials): NodeJS.Process
 export function notifyUserInvalidCredentials(credentialProviderId: CredentialsProviderId): void {
     const getHelp = localize('AWS.generic.message.getHelp', 'Get Help...')
     const viewLogs = localize('AWS.generic.message.viewLogs', 'View Logs...')
+    // getHelp link does not have a corresponding doc page in Cloud9 as of initial launch.
```
Comment (bryceitoc9) [inline src/credentials/credentialsUtilities.ts:31]: Preface with `TODO:` or something else that'll draw attention?

File: src/shared/extensionUtilities.ts [hunk 2]
```diff
@@ -252,10 +312,11 @@ export function getToolkitEnvironmentDetails(): string {
     const vsCodeVersion = vscode.version
     const envDetails = localize(
         'AWS.message.toolkitInfo',
-        'OS:  {0} {1} {2}\nVisual Studio Code Version:  {3}\nAWS Toolkit Version:  {4}\n',
+        'OS:  {0} {1} {2}\n{3} Extension Host Version:  {4}\nAWS Toolkit Version:  {5}\n',
```
Comment (bryceitoc9) [inline src/shared/extensionUtilities.ts:315]: Might be worth changing this to an `ExtensionProperty`

**Comment by justinmk3** [comment]
failed

```
  buildDefaultApiInvokeUrl
    ✓ builds a url
    1) "after each" hook for "builds a url"
  1 passing (37ms)
  1 failing
  1) "after each" hook for "builds a url":
     TypeError: Cannot read property 'restore' of undefined
      at Context.<anonymous> (src/test/lambda/wizards/samDeployWizard.test.ts:655:17)

File: src/shared/sam/debugger/awsSamDebugger.ts [hunk 3]
```diff
@@ -451,6 +458,7 @@ export class SamDebugConfigProvider implements vscode.DebugConfigurationProvider
             },
             awsCredentials: awsCredentials,
             parameterOverrides: parameterOverrideArr,
+            useIkpdb: isCloud9() || !!(config as any).useIkpdb,
```
Comment (bryceitoc9) [inline src/shared/sam/debugger/awsSamDebugger.ts:461]: Swap the order of these arguments? Assuming `useIkpdb` should be prioritized?

**Review by bryceitoc9 — APPROVED** [review]
Couple nits that might make our lives easier in the future, otherwise LGTM!

File: src/shared/sam/debugger/awsSamDebugger.ts (hunk 3 above)
Comment (justinmk3) [reply src/shared/sam/debugger/awsSamDebugger.ts:461]: Only matters in tests

File: src/shared/extensionUtilities.ts (hunk 2 above)
Comment (justinmk3) [reply src/shared/extensionUtilities.ts:315]: ?

Comment (bryceitoc9) [reply src/shared/extensionUtilities.ts:315]: NVM, we are using it (I glanced and saw it as `isCloud9`ing the whole string, not just the long name.

File: src/shared/sam/debugger/awsSamDebugger.ts (hunk 3 above)
Comment (justinmk3) [reply src/shared/sam/debugger/awsSamDebugger.ts:461]: The order does not logically matter.

**Comment by justinmk3** [comment]
Merging this after finding no regressions compared to current `master` branch. 

Found these issues (not regressions):

- Image+python 3.8 issue (cloud9-only): #1442
- Image+API+nodejs issue (present on https://github.com/aws/aws-toolkit-vscode/commit/978f87ba2a79349ae43495b0c98e16d76b8d428f , i.e. before this PR): #1443

**Pull request merged**